appcues/src/main/java/com/appcues/statemachine/states/RenderingStepState.kt

```diff
@@ -0,0 +1,87 @@
+package com.appcues.statemachine.states
+
+import com.appcues.data.model.Experience
+import com.appcues.data.model.StepReference.StepOffset
+import com.appcues.statemachine.Action
+import com.appcues.statemachine.Action.EndExperience
+import com.appcues.statemachine.Action.MoveToStep
+import com.appcues.statemachine.Action.StartStep
+import com.appcues.statemachine.Error.StepError
+import com.appcues.statemachine.State
+import com.appcues.statemachine.Transition
+import com.appcues.statemachine.effects.AwaitDismissEffect
+import com.appcues.statemachine.effects.ContinuationEffect
+
+internal data class RenderingStepState(
+    val experience: Experience,
+    val flatStepIndex: Int,
+    val metadata: Map<String, Any?>
+) : State {
+
+    override val currentExperience: Experience
+        get() = experience
+    override val currentStepIndex: Int
+        get() = flatStepIndex
+
+    override fun take(action: Action): Transition? {
+        return when (action) {
+            is MoveToStep -> if (shouldEndExperience(action)) {
+                toEndingExperience(EndExperience(markComplete = true, destroyed = false))
+            } else {
+                toEndingStep(action)
+            }
+            is EndExperience -> {
+                toEndingExperience(action)
+            }
+            else -> null
+        }
+    }
+
+    private fun shouldEndExperience(action: MoveToStep): Boolean {
+        val isLastStep = flatStepIndex == experience.flatSteps.size - 1
+        val isContinuingToNextStep = action.stepReference is StepOffset && action.stepReference.offset == 1
+
+        return isLastStep && isContinuingToNextStep
+    }
+
+    private fun toEndingStep(action: MoveToStep): Transition {
+        val nextStepIndex = action.stepReference.getIndex(experience, flatStepIndex)
+            ?: return keep(StepError(experience, flatStepIndex, "Step at ${action.stepReference} does not exist"))
+
+        // should never happen but in case next step group lookup returns null we handle it here
+        val stepContainerIndex = experience.groupLookup[nextStepIndex]
+            ?: return keep(StepError(experience, flatStepIndex, "StepContainer for nextStepIndex $nextStepIndex not found"))
```

Comment:
so usually what i've noticed was a lot of "this should ever happen" checks that had no "next state", this was the case for [this](https://github.com/appcues/appcues-android-sdk/blob/main/appcues/src/main/java/com/appcues/statemachine/Transitions.kt#L176C9-L176C24) check too.

and on current implementation when state is null we basically keep it in the same state, the only transition here that changed state to Idling is an exit failure when experience has no steps [here](https://github.com/appcues/appcues-android-sdk/blob/main/appcues/src/main/java/com/appcues/statemachine/Transitions.kt#L49). another "extra check" that is not really needed since we can check flatSteps size when we get to Idling.